packages/@aws-cdk/aws-apprunner/lib/service.ts

```diff
@@ -860,71 +1030,134 @@ export class Service extends cdk.Resource {
     this.serviceStatus = resource.attrStatus;
     this.serviceName = resource.ref;
   }
+
+  /**
+   * This method adds an environment variable to the App Runner service.
+   */
+  public addEnvironment(name: string, value: string) {
+    this.variables.push({ name: name, value: value });
+  }
+
+  /**
+   * This method adds a secret as environment variable to the App Runner service.
+   */
+  public addSecret(name: string, secret: Secret) {
+    if (!this.instanceRole) {
+      this.instanceRole = this.createInstanceRole();
+    }
+    secret.grantRead(this.instanceRole);
+    this.secrets.push({ name: name, value: secret.arn });
+  }
+
+  /**
+   * This method generates an Instance Role. Needed if using secrets and props.instanceRole is undefined
+   * @returns iam.IRole
+   */
+  private createInstanceRole(): iam.IRole {
+    return new iam.Role(this, 'InstanceRole', {
+      assumedBy: new iam.ServicePrincipal('tasks.apprunner.amazonaws.com'),
+      roleName: cdk.PhysicalName.GENERATE_IF_NEEDED,
+    });
+  }
+
+  /**
+   * This method generates an Access Role only when ImageRepositoryType is ECR and props.accessRole is undefined
+   * @returns iam.IRole
+   */
+  private generateDefaultRole(): iam.Role {
+    const accessRole = new iam.Role(this, 'AccessRole', {
+      assumedBy: new iam.ServicePrincipal('build.apprunner.amazonaws.com'),
+    });
+    accessRole.addToPrincipalPolicy(new iam.PolicyStatement({
+      actions: ['ecr:GetAuthorizationToken'],
+      resources: ['*'],
+    }));
+    this.accessRole = accessRole;
+    return accessRole;
+  }
+
+  private getEnvironmentSecrets(): { [key: string]: Secret } | undefined {
+    return this.source.codeRepository?.codeConfiguration.configurationValues?.environmentSecrets ||
+      this.source.imageRepository?.imageConfiguration?.environmentSecrets;
+  }
+
+  private getEnvironmentVariables(): { [key: string]: string } | undefined {
+    return this.source.codeRepository?.codeConfiguration.configurationValues?.environmentVariables ||
+      this.source.codeRepository?.codeConfiguration.configurationValues?.environment ||
+      this.source.imageRepository?.imageConfiguration?.environmentVariables ||
+      this.source.imageRepository?.imageConfiguration?.environment;
```

Comment:
For a `@deprecated` field, we typically also add a validation to make sure you're not passing both (you need to pass either the old or the new one).

Something like:

```ts
if (this.source.codeRepository?.codeConfiguration.configurationValues?.environmentVariables && this.source.codeRepository?.codeConfiguration.configurationValues?.environment) {
  throw new Error('You must supply at most one of \'environmentVariables\' and \'environment\'');
}
```